File: datn_fe/src/app/service/training-room/training-room.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  TrainingRoom,
  TrainingRoomCreateRequest,
  TrainingRoomUpdateRequest,
  TrainingRoomFilterRequest
} from '../../model/training-room.model';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class TrainingRoomService {
  private apiUrl = `${environment.apiUrl}/api/training-room`;

  constructor(private http: HttpClient) { }

  /**
   * Get all training rooms
   */
  getAll(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}`);
  }

  /**
   * Get training room by ID
   */
  getById(id: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${id}`);
  }

  /**
   * Get training room by timeSlots subject ID
   */
  getByTimeSlotsSubjectId(timeSlotsSubjectId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/timeslots-subject/${timeSlotsSubjectId}`);
  }

  /**
   * Get all training rooms for a coach
   */
  getByCoachId(coachId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/coach/${coachId}`);
  }

  /**
   * Get all active training rooms for a coach
   */
  getActiveByCoachId(coachId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/coach/${coachId}/active`);
  }

  /**
   * Get all training rooms for a subject
   */
  getBySubjectId(subjectId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/subject/${subjectId}`);
  }

  /**
   * Get training rooms by status
   */
  getByStatus(status: string): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/status/${status}`);
  }

  /**
   * Search and filter training rooms with pagination
   */
  search(request: TrainingRoomFilterRequest): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/search`, request);
  }

  /**
   * Create a new training room
   */
  create(request: TrainingRoomCreateRequest): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}`, request);
  }

  /**
   * Create training room for a timeSlot subject (auto-generation)
   */
  createForTimeSlotsSubject(timeSlotsSubjectId: number): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/timeslots-subject/${timeSlotsSubjectId}`, {});
  }

  /**
   * Update training room
   */
  update(request: TrainingRoomUpdateRequest): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}`, request);
  }

  /**
   * Delete training room
   */
  delete(id: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${id}`);
  }

  /**
   * Get online training rooms for a user based on their enrolled subjects
   */
  getOnlineRoomsForUser(userId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/user/${userId}`);
  }

  /**
   * Get online training rooms for a user with pagination
   */
  getOnlineRoomsForUserPaginated(userId: number, page: number = 0, size: number = 20): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/user/${userId}/paginated?page=${page}&size=${size}`);
  }

  /**
   * Get online training rooms for a user by specific subject
   */
  getOnlineRoomsForUserBySubject(userId: number, subjectId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/user/${userId}/subject/${subjectId}`);
  }
}

File: datn_be/src/main/java/com/dntn/datn_be/service/TrainingRoomService.java
package com.dntn.datn_be.service;

import com.dntn.datn_be.dto.common.ResponseGlobalDto;
import com.dntn.datn_be.dto.request.TrainingRoomCreateRequest;
import com.dntn.datn_be.dto.request.TrainingRoomFilterRequest;
import com.dntn.datn_be.dto.request.TrainingRoomUpdateRequest;
import com.dntn.datn_be.dto.response.TrainingRoomResponse;
import org.springframework.data.domain.Page;

import java.util.List;

public interface TrainingRoomService {
    
    /**
     * Get all training rooms
     */
    ResponseGlobalDto<List<TrainingRoomResponse>> getAll();
    
    /**
     * Get training room by ID
     */
    ResponseGlobalDto<TrainingRoomResponse> getById(Long id);
    
    /**
     * Get training room by timeSlots subject ID
     */
    ResponseGlobalDto<TrainingRoomResponse> getByTimeSlotsSubjectId(Long timeSlotsSubjectId);
    
    /**
     * Get all training rooms for a coach
     */
    ResponseGlobalDto<List<TrainingRoomResponse>> getByCoachId(Long coachId);
    
    /**
     * Get all active training rooms for a coach
     */
    ResponseGlobalDto<List<TrainingRoomResponse>> getActiveByCoachId(Long coachId);
    
    /**
     * Get all training rooms for a subject
     */
    ResponseGlobalDto<List<TrainingRoomResponse>> getBySubjectId(Long subjectId);
    
    /**
     * Get training rooms by status
     */
    ResponseGlobalDto<List<TrainingRoomResponse>> getByStatus(String status);
    
    /**
     * Search and filter training rooms with pagination
     */
    ResponseGlobalDto<Page<TrainingRoomResponse>> search(TrainingRoomFilterRequest request);
    
    /**
     * Get online training rooms for a user based on their enrolled subjects
     */
    ResponseGlobalDto<List<TrainingRoomResponse>> getOnlineRoomsForUser(Long userId);
    
    /**
     * Get online training rooms for a user with pagination
     */
    ResponseGlobalDto<Page<TrainingRoomResponse>> getOnlineRoomsForUser(Long userId, Integer page, Integer size);
    
    /**
     * Get online training rooms for a user by specific subject
     */
    ResponseGlobalDto<List<TrainingRoomResponse>> getOnlineRoomsForUserBySubject(Long userId, Long subjectId);
    
    /**
     * Create a new training room
     */
    ResponseGlobalDto<TrainingRoomResponse> create(TrainingRoomCreateRequest request);
    
    /**
     * Create training room for a timeSlot subject (auto-generation)
     */
    ResponseGlobalDto<TrainingRoomResponse> createForTimeSlotsSubject(Long timeSlotsSubjectId);
    
    /**
     * Update training room
     */
    ResponseGlobalDto<TrainingRoomResponse> update(TrainingRoomUpdateRequest request);
    
    /**
     * Delete training room
     */
    ResponseGlobalDto<Boolean> delete(Long id);
    
    /**
     * Delete all training rooms for a coach
     */
    ResponseGlobalDto<Boolean> deleteByCoachId(Long coachId);
}

